src/interactivity/win32/ut_interactivity_win32/UiaTextRangeTests.cpp

```diff
@@ -473,6 +473,271 @@ class UiaTextRangeTests
+                    UiaTextRange::MovementIncrement::Backward,
+                    UiaTextRange::MovementDirection::Backward
+                },
+                -1,
+                -1,
+                0u,
+                lastColumnIndex
+            },
+
+            {
+                L"can move forward in a row",
+                {
+                    2, 1,
+                    4, 5,
+                    bottomRow,
+                    firstColumnIndex,
+                    lastColumnIndex,
+                    UiaTextRange::MovementIncrement::Forward,
+                    UiaTextRange::MovementDirection::Forward
+                },
+                5,
+                5,
+                UiaTextRange::_screenInfoRowToEndpoint(_pUiaData, 8),
+                UiaTextRange::_screenInfoRowToEndpoint(_pUiaData, 8) + lastColumnIndex
+            },
+
+            {
+                L"can't move past the last column in the last row",
+                {
+                    bottomRow, lastColumnIndex,
+                    bottomRow, lastColumnIndex,
+                    bottomRow,
+                    firstColumnIndex,
+                    lastColumnIndex,
+                    UiaTextRange::MovementIncrement::Forward,
+                    UiaTextRange::MovementDirection::Forward
+                },
+                5,
+                0,
+                UiaTextRange::_screenInfoRowToEndpoint(_pUiaData, bottomRow),
+                UiaTextRange::_screenInfoRowToEndpoint(_pUiaData, bottomRow) + lastColumnIndex
+            },
+
+            {
+                L"can move to a new row when necessary when moving forward",
+                {
+                    topRow, lastColumnIndex,
+                    topRow, lastColumnIndex,
+                    bottomRow,
+                    firstColumnIndex,
+                    lastColumnIndex,
+                    UiaTextRange::MovementIncrement::Forward,
+                    UiaTextRange::MovementDirection::Forward
+                },
+                5,
+                5,
+                UiaTextRange::_screenInfoRowToEndpoint(_pUiaData, topRow + 5),
+                UiaTextRange::_screenInfoRowToEndpoint(_pUiaData, topRow + 5) + lastColumnIndex
+            },
+
+            {
+                L"can move to a new row when necessary when moving backward",
+                {
+                    topRow + 1, firstColumnIndex,
+                    topRow + 1, lastColumnIndex,
+                    topRow,
+                    lastColumnIndex,
+                    firstColumnIndex,
+                    UiaTextRange::MovementIncrement::Backward,
+                    UiaTextRange::MovementDirection::Backward
+                },
+                -5,
+                -2,
+                0u,
+                UiaTextRange::_screenInfoRowToEndpoint(_pUiaData, topRow) + lastColumnIndex
+            }
+        };
+        // clang-format on
+
+        for (auto data : testData)

```

Comment:
This whole test is the most insane thing I've ever seen but I love it

Reply:
Oh yeah. I'm gonna steal this pattern for keyboard selection too.